File: CPP/Algospot/SORTGAME-2.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <algorithm>

using namespace std;

map<vector<int>, int> cache;
bool calculated[8] = { false };

void preCalculate(int n) {
    if (calculated[n])
        return;

    queue<vector<int>> q;

    vector<int> initial(n, 0);
    for (int i = 0; i < n; i++)
        initial.at(i) = i;

    q.push(initial);
    cache.insert(make_pair(initial, 0));

    while (q.size() > 0) {
        auto front = q.front();
        q.pop();

        auto cost = cache[front];

        for (int i = 0; i < front.size(); i++) {
            for (int j = i + 2; j <= front.size(); j++) {
                reverse(front.begin() + i, front.begin() + j);

                if (cache.count(front) == 0) {
                    cache.insert(make_pair(front, cost + 1));
                    q.push(front);
                }

                reverse(front.begin() + i, front.begin() + j);
            }
        }
    }

    calculated[n] = true;
}

void solve() {
    int N;
    cin >> N;

    vector<int> numbers(N, 0);
    vector<int> converted(N, 0);

    for (int i = 0; i < N; i++) {
        int val; cin >> val;
        numbers.at(i) = val;
    }

    for (int i = 0; i < N; i++) {
        int less = 0;
        for (int j = 0; j < N; j++)
            if (numbers.at(j) < numbers.at(i))
                less++;
        
        converted.at(i) = less;
    }

    preCalculate(N);
    int result = cache.at(converted);

    cout << result << "\n";
}

int main() {
    int C;
    cin >> C;

    for (int i = 0; i < C; i++)
        solve();

    return 0;
}

File: Python/algospot/SORTGAME-2.py
from collections import deque
# li 라는 배열을, start index 부터 end index 까지 reverse
def reverse(li, start, end):
    lo = start
    hi = end
    while (lo < hi):
        temp = li[hi]
        li[hi] = li[lo]
        li[lo] = temp
        lo += 1; hi -= 1

memo = dict()

def preCalculate(n): # int -> void
    initial = tuple([i for i in range(1, n + 1)])

    queue = deque([[0, initial]])

    while len(queue) > 0:
        curr = queue.popleft()
        
        if curr[1] in memo:
            continue
        else:
            memo[curr[1]] = curr[0]

        li = list(curr[1])
        for i in range(0, n):
            for j in range(i+1, n):
                if i != j:
                    reverse(li, i, j)
                    queue.append([curr[0] + 1, tuple(li)])
                    reverse(li, i, j)

def normalizeNumbers(li):
    result = []
    for i in range(len(li)):
        greaterCount = 1
        for j in range(len(li)):
            if li[i] > li[j]:
                greaterCount += 1
        
        result.append(greaterCount)

    return result

# main
C = int(input())

for i in range(C):
    N = int(input())
    numbers = list(map(int, input().split()))
    numbers = normalizeNumbers(numbers)

    preCalculate(N)
    print(memo[tuple(numbers)])